### ISSUE TITLE
[no-unnecessary-type-assertion] False positive when asserting a subtype of Element

### ISSUE THREAD

**Issue by mitchchn** [typescript-eslint/typescript-eslint#391]
**Repro**

```TS
const el = document.querySelector(`li.focus`) as HTMLLIElement | null;
if (el) {
   el.focus();
 }
```

**Expected Result**

This type assertion is necessary. It narrows `Element` to `HTMLLIElement`, allowing `focus` to be called on the element after the null check.

**Actual Result**

It is flagged as unnecessary:

```
warning  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
```

And if I use `--fix`, the assertion is removed, breaking the build.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.5.0` |
| `@typescript-eslint/parser`        | `1.5.0` |
| `TypeScript`                       | `3.4.1` |
| `ESLint`                           | `5.16.0` |

**Comment by bradzacher** [comment]
Couldn't you instead do
```TS
const el: HTMLLIElement | null = document.querySelector('li.focus'); 

if (el) {
  el.focus();
}
```
[repl](https://www.typescriptlang.org/play/#src=const%20el%3A%20HTMLLIElement%20%7C%20null%20%3D%20document.querySelector('li.focus')%3B%20%0A%0Aif%20(el)%20%7B%0A%20%20el.focus()%3B%0A%7D%0A)

The intention of the rule is to stop you from using the `as` operator unless it is absolutely necessary.
In this case, it is not absolutely necessary to use the `as` operator, as you can type the variable instead.

**Comment by mitchchn** [comment]
Thanks. I guess my beef is more with TypeScript than with ESLint, because it feels wrong to use a type annotation with `querySelector` when the result isn't known at compile time. But I see that the typing of `querySelector` permits it:

`querySelector<E extends Element = Element>(selectors: string): E | null;`

I'd still prefer to use an annotation but agree that it's not _necessary_ and that this issue can be closed.

**Comment by bradzacher** [comment]
You could also save the type annotation and convert it to a runtime check if you prefer that:

```TS
const el = document.querySelector('li.focus'); 

if (el instanceof HTMLLIElement) {
  el.focus();
}
```
